Feature/recurring rent transactions

Jacu93 opened the pull request:
It's been a while since last pull request and for that reason I'd like to close this part already :D Just as a quick reminder this branch introduces recurring monthly payments without need of an actual scheduler. Main feature is balance presentation on the dashboard. You can see last 2 months balance for each tenant. Balance for each month is the total amount from transactions split among all tenants plus rent value (also split). Everytime new rent value is defined it will be charged next month.

I think that all improvements I still have to do are listed here:
https://docs.google.com/document/d/14g8G9dr_y50B8FHUGLYVBBThs0iBfc_VN79tngBzc40/edit
but I think it will be the best to move them onto new branch. Please share your opinion.

Jacu93 committed:
Implementing JwtData annotation to simplify header extraction process

Jacu93 committed:
Removing constructor from JwtData. Logging undefined exceptions to console.

Jacu93 committed:
Added update rent value method to the ApartmentController

Jacu93 committed:
Moving balance management to TransactionsService

Jacu93 committed:
Balance related tests moved (and refactored) from ApartmentServiceTest to TransactionsServiceTest

Jacu93 committed:
Script for loading apartment dropdowns is now separated to avoid code duplication. You can now add new rent value from administration page level.

Jacu93 committed:
Adding new method getLastMonthsBalance to the controller. It will always return balance of all tenants for selected apartment for the last 2 months. Token is now validated in every API request.

Jacu93 committed:
Dashboard balance representation

Jacu93 committed:
Visual dashboard corrections. Email of each tenant is now presented.

Jacu93 committed:
Correction in API. When updating rent in December there would be an exception because of wrong month calculation. Now corrected.
Starting implementation of rent checking mechanism for the last two months when in dashboard.

Jacu93 committed:
Further implementation of rent checking mechanism for the last two months when in dashboard. Added new API method to get balance for last 2 months. Extending homePage.js to handle balance presentation.

Jacu93 committed:
Corrections in homePage.js in order to present balance based on rent value if there are no transactions for desired month. Removing debug console logging.

Jacu93 committed:
Further corrections fo homePage.js. Balance is now presented as summed up value instead of merged string. Removed initial value from transactions.html.

piekaa commented on src/main/java/com/piekoszek/nowaksharedrent/transactions/TransactionsController.java:
```diff
@@ -23,7 +23,7 @@
+    ResponseEntity<Object> getTransactionsFromMonth(@PathVariable("month") int month, @PathVariable("year") int year, @PathVariable("apartmentId") String apartmentId, @Jwt JwtData jwtData) {
         Transactions transactions = transactionsService.getTransactionsFromMonth(month, year, apartmentId);
         if (transactions == null) {
             return new ResponseEntity<>(new MessageResponse("No transactions with given id found."), HttpStatus.BAD_REQUEST);
```
I know it was not changed, but NOT_FOUND fits better.

piekaa commented on src/main/java/com/piekoszek/nowaksharedrent/transactions/TransactionsServiceImpl.java:
```diff
@@ -75,4 +79,21 @@ public Transactions getTransactionsFromMonth(int month, int year, String apartme
     }
+
+    @Override
+    public Map<String, Set<Payer>> getLastMonthsBalance(String apartmentId) {
```
rename to getLastTwoMonthsBalance or parametrize

piekaa commented on src/main/java/com/piekoszek/nowaksharedrent/apartment/ApartmentServiceImpl.java:
```diff
@@ -53,16 +58,24 @@ public boolean hasTenant(Apartment apartment, String email) {
-        Apartment apartment = apartmentRepository.findById(apartmentId);
-        apartment.updateBalance(value);
+        Calendar currDate = timeService.currentDateAndTime();
+        if (currDate.get(Calendar.MONTH) == Calendar.DECEMBER) {
```
I think without this if - else it would be better (rent would be set for current month).

Jacu93 replied:
Rent values are stored in order as you add them anyway but in the feature I might add just the timestamp so it is certain which value is the last. I have to make sure that beginning of the month doesn't matter in the rest of the code. For now I will just consider your suggestion :)

piekaa commented on src/main/resources/static/scripts/homePage.js:
```diff
@@ -88,28 +88,51 @@ function loadRoommates(ApartmentId) {
+            let borderDate = new Date();
+            let rentsArrayLength = rentsArray.length;
+            tenantsArrayLength = tenantsArray.length;
+            while (rentsArrayLength>0) {
```
`for(i = rentsArrayLength ; i >=0 ; i--)`

piekaa commented on src/main/resources/static/scripts/homePage.js:
```diff
@@ -88,28 +88,51 @@ function loadRoommates(ApartmentId) {
+                if (borderDate.getFullYear() < currentDate.getFullYear() || 
+                    (borderDate.getMonth() < currentDate.getMonth() && borderDate.getFullYear() == currentDate.getFullYear())) {
+                    lastMonthRent = rentsArray[rentsArrayLength].value;
+                    rentsArrayLength = 0;
```
`break;`

piekaa commented on src/main/resources/static/scripts/transactionsPage.js:
```diff
@@ -2,25 +2,26 @@ var transactionTypesEnum = null;
         getTransactionTypesEnum(loadTransactionTypesDropdown);
-        loadTransactions();
         let currDate = new Date();
         $('.datepicker').datepicker({
```
assign $('.datepicker') to variable and reuse

piekaa reviewed (changes requested):
Sure, move improvements to another branch. Let's do small steps.

piekaa commented:
Also remember to change loop in then: in TransactionsServiceTest

Jacu93 committed:
Pull request corrections #1
getTransactionsFromMonth change BAD_REQUEST -> NOT_FOUND
getLastMonthsBalance -> getLastTwoMonthsBalance
new rent value is used in the current month
while -> for with break in homePage.js (110)
removed duplicated jQuery selector for datepicker

Jacu93 committed:
Pull request corrections #2
for loops in TransactionsServiceTest file replaced with payers.each method

Jacu93 commented:
Corrections done as requested

piekaa commented on src/main/resources/static/scripts/homePage.js:
```diff
@@ -107,17 +107,19 @@ function loadRoommates(ApartmentId) {
-            while (rentsArrayLength>0) {
-                rentsArrayLength--;
-                borderDate = new Date(rentsArray[rentsArrayLength].borderDate);
+            for(i = rentsArrayLength ; i >=0 ; i--) {  
```
could be let i, but nvm, just tip for future

piekaa reviewed (approved):
🥇
